source/common/event/dispatcher_impl.cc

```diff
@@ -162,8 +162,14 @@ void DispatcherImpl::run(RunType type) {
 }
 
 void DispatcherImpl::runPostCallbacks() {
-  std::function<void()> callback;
   while (true) {
+    // It is important that this declaration is inside the body of the loop so that the callback is
+    // destructed while post_lock_ is not held. If callback is declared outside the loop and reused
+    // for each iteration, the previous iteration's callback is destructed when callback is
+    // re-assigned, which happens while holding the lock. This can lead to a deadlock (via
+    // recursive mutex acquisition) if destroying the callback runs a destructor, which through some
+    // callstack calls post() on this dispatcher.
+    std::function<void()> callback;
```

Comment:
though not directly related to this fix, since you're here, how much is the cost by copying callback function here? could this be more efficient by swapping the whole list from `post_callbacks_` and run them, instead of locking for each callbacks?

Reply:
Yes that would be more efficient to just move the whole list and lock once. I would be fine with changing the code to do that also.

Reply:
I assume you'd still need the loop to collect callbacks that get posted after the loop starts.

Reply:
It wouldn't be explicitly necessary. When the array goes from 0 -> 1 a new message would be posted to run callbacks again. A single move would work.

Reply:
I think for this change I would rather keep it as a minimal bug fix. Someone can go through later to make it faster by swapping the list out.